src/subtype.c

```diff
@@ -0,0 +1,569 @@
+// This file is a part of Julia. License is MIT: http://julialang.org/license
+
+/*
+  subtyping predicate
+*/
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#ifdef _OS_WINDOWS_
+#include <malloc.h>
+#endif
+#include "julia.h"
+#include "julia_internal.h"
+
+typedef struct {
+    int depth;
+    int8_t more;
+    int stacksize;
+    uint32_t stack[10];
+} jl_unionstate_t;
+
+typedef struct _varbinding {
+    jl_tvar_t *var;
+    jl_value_t *lb;
+    jl_value_t *ub;
+    jl_tvar_t *concretevar;
+    int8_t right;
+    int8_t occurs_inv;
+    int8_t occurs_cov;
+    int8_t concrete;
+    int depth0;
+    struct _varbinding *prev;
+} jl_varbinding_t;
+
+typedef struct {
+    jl_varbinding_t *vars;
+    jl_unionstate_t Lunions;
+    jl_unionstate_t Runions;
+    jl_value_t **envout;
+    int envsz;
+    int envidx;
+    int invdepth;
+} jl_stenv_t;
+
+// state manipulation utilities
+
+static jl_varbinding_t *lookup(jl_stenv_t *e, jl_tvar_t *v)
+{
+    jl_varbinding_t *b = e->vars;
+    while (b != NULL) {
+        if (b->var == v) return b;
+        b = b->prev;
+    }
+    return b;
+}
+
+static int statestack_get(jl_unionstate_t *st, int i)
+{
+    assert(i < st->stacksize);
+    return (st->stack[i>>5] & (1<<(i&31))) != 0;
```

Comment:
Just leave as a TODO list for when you get to it---as an author it's not always obvious what needs commenting. For me, a first order priority is to make sure I can still hack on this sometime in the future.